tsdb/importer/importer.go

```diff
@@ -0,0 +1,602 @@
+		return bins
+	}
+
+	bin := make([]*newBlockMeta, 0)
+
+	start := blocks[0].mint
+	blockIdx := 0
+Outer:
+	for dIdx, duration := range durations {
+		for bIdx, block := range blocks[blockIdx:] {
+			blockIdx = bIdx
+			if block.maxt-block.mint >= duration {
+				if block.mint <= start {
+					bin = append(bin, block)
+					bins = append(bins, bin)
+				} else {
+					bins = append(bins, bin)
+					bins = append(bins, []*newBlockMeta{block})
+				}
+				bin = []*newBlockMeta{}
+				start = block.maxt
+				if dIdx == len(durations)-1 {
+					continue
+				} else {
+					blockIdx++
+					continue Outer
+				}
+			}
+			if block.mint-start < duration && block.maxt-start < duration {
+				// If current block is within the duration window, place it in the current bin.
+				bin = append(bin, block)
+				blockIdx++
+			} else {
+				// Else create a new block, starting with this block.
+				bins = append(bins, bin)
+				bin = []*newBlockMeta{block}
+				start = block.mint
+				if dIdx < len(durations)-1 {
+					blockIdx++
+					continue Outer
+				}
+			}
+		}
+	}
+	bins = append(bins, bin)
+	return bins
+}
+
+// binSamples divvies the samples into bins corresponding to the given block.
+// If an aligned block is given to it, then a single bin is created,
+// else, it divides the samples into blocks of duration.
+// It returns the binned samples, and the time limits for each bin.
+func binSamples(samples []*tsdb.MetricSample, meta *newBlockMeta, duration int64) ([][]*tsdb.MetricSample, []blockTimestampPair) {
+	findBlock := func(ts int64, blocks []blockTimestampPair) int {
+		for idx, block := range blocks {
+			if block.start <= ts && ts < block.end {
+				return idx
+			}
+		}
+		return -1
+	}
+	bins := make([][]*tsdb.MetricSample, 0)
+	times := make([]blockTimestampPair, 0)
+	if len(samples) == 0 {
+		return bins, times
+	}
+	if meta.isAligned {
+		bins = append(bins, samples)
+		times = []blockTimestampPair{{start: meta.mint, end: meta.maxt}}
+	} else {
+		timeTranches := makeRange(meta.mint, meta.maxt, duration)
+		bins = make([][]*tsdb.MetricSample, len(timeTranches))
+		binIdx := -1
+		for _, sample := range samples {
+			ts := sample.Timestamp
+			if binIdx == -1 {
+				binIdx = findBlock(ts, timeTranches)
+			}
+			block := timeTranches[binIdx]
+			if block.start <= ts && ts < block.end {
+				bins[binIdx] = append(bins[binIdx], sample)
+			} else {
+				binIdx = findBlock(ts, timeTranches)
+				bins[binIdx] = append(bins[binIdx], sample)
+			}
+		}
+		times = timeTranches
+	}
+	return bins, times
+}
+
+// constructBlockMetadata gives us the new blocks' metadatas.
+func constructBlockMetadata(path string) ([]*newBlockMeta, error) {
+	dbMint, dbMaxt := int64(math.MinInt64), int64(math.MaxInt64)
+
+	// If we try to open a regular RW handle on an active TSDB instance,
+	// it will fail. Hence, we open a RO handle.
+	db, err := tsdb.OpenDBReadOnly(path, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+
+	blocks, err := db.Blocks()
```

Comment:
I don't think we can rely on being in Prometheus DB path. Think about cases when you run Promethues on remote machines